resources/js/features/game-list/hooks/useTableSync.ts

```diff
@@ -1,25 +1,61 @@
-import type { ColumnFiltersState, PaginationState, SortingState } from '@tanstack/react-table';
-import { useUpdateEffect } from 'react-use';
+import type {
+  ColumnFiltersState,
+  PaginationState,
+  SortingState,
+  TableState,
+  VisibilityState,
+} from '@tanstack/react-table';
+import { useCookie, useUpdateEffect } from 'react-use';
+
+import { usePageProps } from '@/common/hooks/usePageProps';
 
 interface UseAutoUpdatingQueryParamsProps {
   columnFilters: ColumnFiltersState;
+  columnVisibility: VisibilityState;
   defaultFilters: ColumnFiltersState;
   pagination: PaginationState;
   sorting: SortingState;
 
   defaultPageSize?: number;
+  isUserPersistenceEnabled?: boolean;
 }
 
 /**
- * This hook is designed to keep the URL query params in sync with the table state.
+ * This hook is designed to keep the URL query params and
+ * user's persistence cookie in sync with the table state.
  */
-export function useAutoUpdatingQueryParams({
+export function useTableSync({
   columnFilters,
+  columnVisibility,
   pagination,
   sorting,
   defaultFilters = [],
   defaultPageSize = 25,
+  isUserPersistenceEnabled = false,
 }: UseAutoUpdatingQueryParamsProps) {
+  const { persistenceCookieName } = usePageProps<{ persistenceCookieName: string }>();
+
+  const [cookie, setCookie, deleteCookie] = useCookie(persistenceCookieName);
+
+  useUpdateEffect(() => {
+    if (isUserPersistenceEnabled) {
+      // Don't persist filtering by title.
+      const persistedFilters = columnFilters.filter((filter) => filter.id !== 'title');
+
+      const tableState: Partial<TableState> = {
+        columnVisibility,
+        sorting,
+        columnFilters: persistedFilters,
+        pagination: { ...pagination, pageIndex: 0 }, // don't persist the page index
+      };
+
+      setCookie(JSON.stringify(tableState), { expires: 180 }); // 180 day (6 month) expiry
```

Comment:
JSON is a very inefficient storage mechanism for cookies because of all the escaping:
![image](https://github.com/user-attachments/assets/d875d5ee-2642-477c-9eb5-4506782163a7)

But I don't have any better solutions. Perhaps we could restructure the data to eliminate most of the structure?
```
{"columnVisibility":{"hasActiveOrInReviewClaims":false,"lastUpdated":false,"numUnresolvedTickets":false,"numVisibleLeaderboards":false,"playersTotal":false,"progress":true,"retroRatio":false},"sorting":[{"id":"title","desc":false}],"columnFilters":[{"id":"achievementsPublished","value":["either"]},{"id":"system","value":["27"]},{"id":"progress","value":["unstarted"]}],"pagination":{"pageIndex":0,"pageSize":25}};
```
to
```
{"columnVisibility":"!hasActiveOrInReviewClaims|!lastUpdated|!numUnresolvedTickets|!numVisibleLeaderboards|!playersTotal|progress|!retroRatio","sort":"title","columnFilters":[{"achievementsPublished":["either"]},{"system":["27"]},{"progress":["unstarted"]}],"pageIndex":0,"pageSize":25};
```
Although, any sort of custom mutation introduces a point of failure. Is there some way to better filter out default values?

Reply:
base64 encoding should generate about the same rate of compression. It's a fixed 33% bloat, whereas JSON encoding is 200% bloat, but only for for the structure characters, so if the keys and data compose 83% of the message, you end up with the same 33% bloat.

I did actually try that the other day, and my cookie went from something like 570 characters (escaped) to 550 characters (base64 encoded).

I'd be fine if you feel more comfortable using the escaped methodology.

Reply:
Undid `btoa()` in latest. If we'd like to pursue a compression algorithm in the future, here's a few benchmarks/comparisons for algorithms that are design to compress JSON: https://senryoku.github.io/smol-string/